ouroboros-consensus/src/Ouroboros/Consensus/Util/ResourceRegistry.hs

```diff
@@ -1037,6 +1073,23 @@ releaseAllHelper rr context releaser = mask_ $ do
       Left closed -> throwRegistryClosed rr context closed
       Right keys  -> releaseResources rr keys releaser
 
+-- | Convenience combinator for allocating a resource and then doing
```

Comment:
I was wavering whether the release-reference is necessary. Generally, the components of the ChainDB should not be older than the ChainDB itself in the registry. But that's only a concern if the components end up in the Consensus registry. If they'll never be in that registry (only transitively via the ChainDB registry), then we can just use temp registries to hold things until the ChainDB itself is fully initialized and added to the Consensus registry.

Reply:
Javier's `Revamp` commit started along these lines. Here's some thoughts in response to what he's got in there so far.

(1) The `allocateTemp` function is incompatible with encapsulation. Specifically, if the `st` inhabitant is a closure that contains the resource, then we can't muster the third argument here.

```
-- | Allocate a resource in a temporary registry until it has been transferred
-- to the final state @st@. See 'runWithTempRegistry' for more details.
allocateTemp
  :: (IOLike m, HasCallStack)
  => m a
     -- ^ Allocate the resource
  -> (a -> m Bool)
     -- ^ Release the resource, return 'True' when the resource was actually
     -- released, return 'False' when the resource was already released.
     --
     -- Note that it is safe to always return 'True' when unsure.
  -> (st -> a -> Bool)
     -- ^ Check whether the resource is in the given state
  -> WithTempRegistry st m a
```

Specifically, your current code

```
    resourcesFromChainDB :: ChainDbEnv m blk -> Resources m blk
    resourcesFromChainDB = undefined -- TODO @js
```

is unimplementable unless we change `ChainDbEnv` to include not only the `ImmutableDB` interface closure but also the `ImmutableDBEnv` and so on. That seems undesirable, and I'm not sure the `allocateTemp` run-time check is worth that leakage. (If we must, then we could introduce a newtype to make it extremely clear that that leakage is only there for one reason, so that code review would relatively-easily catch abuses of it. Maybe there are other module-based or rank2-based approaches to mitigate the leakage.)

(2) Therefore, I suggest that we start with a PR that just passes `\_ _ -> True` to the new uses of `allocateTemp` and uses eg `st ~ ()`. We lose the assurance of the run-time check, but I think that's a good place to start (even possibly merging it and taking the lack of run-time check as tech debt).

However, this might not be so simple, since some of the existing `openDBInternal` code seems to _properly_ use `allocateTemp` already (eg `Ouroboros.Consensus.Storage.ImmutableDB.Impl.openDBInternal` on the `master` branch).

(3) Supposing we do succumb to the additional leakage, then I think we'll need to relax the `WithTempRegistry` type a bit.

It is currently:

```
newtype WithTempRegistry st m a = WithTempRegistry { unWithTempRegistry :: ReaderT (TempRegistry st m) m a }

data TempRegistry st m = TempRegistry {
      tempResourceRegistry :: !(ResourceRegistry m)
    , tempTransferredTo    :: !(StrictTVar m (TransferredTo st))        -- ^ Used as a @Writer@.
    }

runWithTempRegistry :: (IOLike m, HasCallStack) => WithTempRegistry st m (a, st) -> m a
```

We could instead do something like

```
data TempRegistry st m = forall largest_st. TempRegistry {
      tempResourceRegistry :: !(ResourceRegistry m)
    , tempProjector        :: !(largest_st -> st)
    , tempTransferredTo    :: !(StrictTVar m (TransferredTo largest_st))        -- ^ Used as a @Writer@.
    }

runWithTempRegistry = .... tempProjector = id ...

weakenWithTempRegistry :: (st' -> st) -> WithTempRegistry st m a -> WithTempRegistry st' m a
weakenWithTempRegistry prj = ... tempProjector = prj . tempProjector ...
```

Such a `weakenWithTempRegistry` supersedes your PR's current `decorateWithTempRegistry` function, I think. And thereby avoids the proximal `undefined`s etc.

Reply:
Hmmm.

```
    , tempProjector        :: !(forall a. (st -> a -> Bool) -> largest_st -> a -> Bool)

weakenWithTempRegistryAns ::
     ((forall b. (st -> b -> Bool) -> largest_st -> b -> Bool))
  -> WithTempRegistry st m a
  -> WithTempRegistry st' m a
```

Such a `weakenWithTempRegistryAns` would enable the (2) suggestion in my previous comment above, because it would let us override all of the checks in the weakened `WithTempRegistry` to just all return `True`.

Reply:
I'm closing this in favor of a new Conversation.